[addons] add help for load of shared library and allow use of own addon temp folder

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This is a part to simplify the load of shared library functions on a
addon.

This part is mainly needed for audio decoder addons.

Further add it a new addon callback function to get a temporary folder for addon itself. This becomes used e.g. on audio decoders to copy temporary a used dll to there.
Will make this way to ensure that temporary things are really erased again and there is no flooding on the disk.

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented:
@notspiff Thats the part what I testet on your addon

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/DllHelper.h:
```diff
@@ -0,0 +1,129 @@
+  template <typename T>
+  bool RegisterSymbol(T& functionPtr, const char* strFunctionPtr)
+  {
+    functionPtr = (T)dlsym(m_dll, strFunctionPtr);
```
c++ style cast instead of c style cast, please.

ksooo reviewed (commented):
Looking good. But where do you want to use this?

AlwinEsch commented:
See there the second commit https://github.com/notspiff/audiodecoder.timidity/pull/17

ksooo commented:
Okay. I assume you will remove Dllhelper.h from notspiff/audiodecoder.timidity#17 ?

AlwinEsch commented:
Yes was only to ask him.

There becomes also another commit added here to become a temporary folder for addon. On test it use the addon user folder, but not like to do there.

I add a function where a addon can get a individual folder where he can use for temporary parts, if now the addon becomes closed, becomes this deleted. Is to prevent in case of errors a flooded harddisk.

AlwinEsch committed:
[addons] add tool header to help with load of shared library

This is a part to simplify the load of shared library functions on a
addon.

This part is mainly needed for audio decoder addons.

AlwinEsch commented:
Is updated

ksooo commented on xbmc/addons/binary-addons/AddonDll.cpp:
```diff
@@ -323,6 +323,11 @@ void CAddonDll::Destroy()
   if (m_pDll)
   {
+    /* If temporary directory was used from addon delete them */
+    std::string tempPath = URIUtils::AddFileToFolder("special://temp/addons", ID());
```
this could be a "const std::string".

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -239,4 +243,20 @@ void Interface_General::get_md5(void* kodiBase, const char* text, char* md5)
+    return nullptr;
+  }
+
+  std::string tempPath = URIUtils::AddFileToFolder("special://temp/addons", addon->ID());
```
this could be a "const std::string".

ksooo reviewed (approved):
you should rename the PR to reflect that it now also contains the temp path access function.

AlwinEsch committed:
[addons] add possibility to get a own temporary addon folder

This is to have on addon a temporary folder, This becomes deleted from
Kodi after the addon is stopped.
